pkg/suggestion/nasrl_service.py

```diff
@@ -63,24 +79,27 @@ def setup_controller(self, request):
                 optim_algo=ctrl_param['optimizer'],
                 skip_target=ctrl_param['skip-target'],
                 skip_weight=ctrl_param['skip-weight'],
-                name="Ctrl_"+request.study_id)
+                name="Ctrl_"+request.study_id,
+                logger=self.logger)
 
-            self.controllers.build_trainer()
+            self.controllers[studyID].build_trainer()
 
-        self.logger.info("Suggestion for StudyJob {} has been initialized.".format(request.study_id))
+        self.logger.info("Suggestion for StudyJob {} (ID: {}) has been initialized.\n".format(self.study_names[studyID], studyID))
 
     def GetSuggestions(self, request, context):
+        studyID = request.study_id
+
         if request.study_id not in self.registered_studies:
             self.setup_controller(request)
-            self.is_first_run = True
-            self.registered_studies.append(request.study_id)
+            self.is_first_run[studyID] = True
```

Comment:
all self.dict will always increase, and nasrl-suggestion service will consume all memory (I think before that OOM killer will kill the pod)

Reply:
@hougangliu Is it possible to get this value: https://github.com/kubeflow/katib/blob/master/pkg/api/operators/apis/studyjob/v1alpha1/studyjob_types.go#L34 
inside Suggestion, right now?

Reply:
RequestCount is the only information to determine whether a job has ended or not. If we cannot get that, we could not delete the StudyJob information.

Reply:
In fact, a studyjob may stop calling GetSuggestion from suggestion service when: 1. changed to `failed`; 2. changed to `completed` (exceed RequestCount or exceed goal).
I think we had to add a new api like `StopSuggestion`  in above two conditions to make suggestionService to clear something. @YujiOshima @richardsliu @johnugeorge WDYT?

Reply:
@hougangliu @YujiOshima Should we ignore this parameter at this time? 
It will work, but we can't clean dictionary with studyIDs and information about it.
After we add RequestCount parameter to the API, we can modify this functionality.

Reply:
Please trace cleaning unused controller by an issue. thanks!